File: src/main/java/edu/kit/iti/formal/stvs/view/spec/variables/clipboard/Json.java
package edu.kit.iti.formal.stvs.view.spec.variables.clipboard;

import com.google.gson.Gson;
import edu.kit.iti.formal.stvs.model.common.FreeVariable;

import java.util.List;
import java.util.stream.Collectors;

/**
 * This class handles the conversion from a lis of {@link FreeVariable} to JSON and vice versa.
 *
 * @author Philipp
 */
public class Json {
  private static final Gson GSON = new Gson();

  private Json() {}

  public static class FreeVarSelection {
    public List<FreeVar> selected;
  }

  public static class FreeVar {
    public String name;
    public String type;
    public String defaultval;
  }

  /**
   * Generates JSON from variables.
   * @param freeVariables variables to convert
   * @return Stringified JSON version of variables
   */
  public static String stringFromRealFreeVariables(List<FreeVariable> freeVariables) {
    return GSON.toJson(fromRealFreeVariables(freeVariables), FreeVarSelection.class);
  }

  /**
   * Generates variables from JSON.
   * @param input Stringified JSON version of variables
   * @return restored variables
   */
  public static List<FreeVariable> stringToRealFreeVariables(String input) {
    return toRealFreeVariables(GSON.fromJson(input, FreeVarSelection.class));
  }

  /**
   * Generates a stringifyable {@link FreeVarSelection} from a list of {@link FreeVariable}.
   * @param freeVariables variables to convert
   * @return converted selection
   */
  public static FreeVarSelection fromRealFreeVariables(List<FreeVariable> freeVariables) {
    List<FreeVar> vars = freeVariables.stream().map(freeVariable -> {
      FreeVar var = new FreeVar();
      var.name = freeVariable.getName();
      var.type = freeVariable.getType();
      var.defaultval = freeVariable.getDefaultValue();
      return var;
    }).collect(Collectors.toList());
    FreeVarSelection selection = new FreeVarSelection();
    selection.selected = vars;
    return selection;
  }

  /**
   * Generates a list of {@link FreeVariable} from the stringifyable class {@link FreeVarSelection}.
   *
   * @param selection stringifyable selection
   * @return list of variables
   */
  public static List<FreeVariable> toRealFreeVariables(FreeVarSelection selection) {
    return selection.selected.stream()
        .map(freeVar -> new FreeVariable(freeVar.name, freeVar.type, freeVar.defaultval))
        .collect(Collectors.toList());
  }
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/view/spec/variables/clipboard/Json.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.view.spec.variables.clipboard

import edu.kit.iti.formal.stvs.model.common.FreeVariable

/**
 * This class handles the conversion from a list of [FreeVariable] to JSON and vice versa.
 *
 * @author Philipp
 */
object Json {
    var json = Json

    /**
     * Generates JSON from variables.
     * @param freeVariables variables to convert
     * @return Stringified JSON version of variables
     */
    fun stringFromRealFreeVariables(freeVariables: List<FreeVariable>): String {
        return "" // return json.encodeToString(fromRealFreeVariables(freeVariables))
    }

    /**
     * Generates variables from JSON.
     * @param input Stringified JSON version of variables
     * @return restored variables
     */
    fun stringToRealFreeVariables(input: String?): List<FreeVariable> {
        // if (input == null)
        return listOf()
        // return toRealFreeVariables(json.decodeFromString<FreeVarSelection>(input))
    }

    /**
     * Generates a stringifyable [FreeVarSelection] from a list of [FreeVariable].
     * @param freeVariables variables to convert
     * @return converted selection
     */
    fun fromRealFreeVariables(freeVariables: List<FreeVariable>): FreeVarSelection {
        val vars = freeVariables.map { FreeVar(it.name, it.type, it.constraint) }
        return FreeVarSelection(vars)
    }

    /**
     * Generates a list of [FreeVariable] from the stringifyable class [FreeVarSelection].
     *
     * @param selection stringifyable selection
     * @return list of variables
     */
    fun toRealFreeVariables(selection: FreeVarSelection): List<FreeVariable> =
        selection.selected.map { FreeVariable(it.name, it.type, it.defaultval) }

    data class FreeVarSelection(var selected: List<FreeVar> = listOf())

    data class FreeVar(var name: String? = null, var type: String? = null, var defaultval: String? = null)
}